raft.c

```diff
@@ -2346,6 +2361,89 @@ void applyShardGroupChange(RedisRaftCtx *rr, raft_entry_t *entry)
     }
 }
 
+void replaceShardGroups(RedisRaftCtx *rr, raft_entry_t *entry)
+{
+    // 1. reset sharding info
+    ShardingInfo *si = rr->sharding_info;
+
+    ShardGroup *local = NULL;
+    if (si->shard_group_map != NULL) {
+        RedisModuleDictIter *iter = RedisModule_DictIteratorStartC(si->shard_group_map, "^", NULL, 0);
+
+        char *key;
+        size_t key_len;
+        ShardGroup *data;
+
+        while ((key = RedisModule_DictNextC(iter, &key_len, (void **) &data)) != NULL) {
+            /* local shardgroup will not have a filled in id */
+            if (*data->id != 0) {
+                RedisModule_Free(data);
+            } else {
+                local = data;
+            }
+        }
+        RedisModule_DictIteratorStop(iter);
+        RedisModule_FreeDict(rr->ctx, si->shard_group_map);
+        si->shard_group_map = NULL;
+    }
+
+    /* if we didn't identify the local shard group, we have a problem */
+    RedisModule_Assert(local != NULL);
```

Comment:
So, this should go into "if check" above to set `local`. Do we always have `si->shard_group_map`? Can it be null if "replace" is the first command. I'm not sure, just asking, if user mistake can trigger an assert.

Reply:
yes, we always have a shardgroup.  its created at node creation time (even before cluster init / join)

Reply:
Okay, nice. So, probably null checks for it are leftover in the codebase, maybe better to remove them to avoid confusion.